### ISSUE TITLE
ipc socket not created after restart

### ISSUE THREAD

**Issue by lahwaacz** [qutebrowser/qutebrowser#3210]
Somehow I got into a state where the ipc socket is not created in the runtime directory after `:restart`. This is obviously bad because then running `qutebrowser some.webpage.org` starts a new instance with the same session (basically all open tabs cloned into the new instance and one extra tab for `some.webpage.org`).

I tried to reproduce it with `--temp-basedir` and when there is still only one tab, the socket is created even after `:restart`. But if I `:tab-clone` the DuckDuckGo tab multiple times to have about 20 tabs open, the socket is missing after `:restart`. I noticed that on each restart, there is this warning:
```
WARNING: /usr/lib/python3.6/subprocess.py:761: ResourceWarning: subprocess 4248 is still running
  ResourceWarning, source=self)
```
In the 20-tabs case the original window takes much longer to disappear and sometimes there are even 2 windows next to each other for a moment. When I close the 20-tabs instance after restart, I get this:
```
ERROR: Could not remove crash log!
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/qutebrowser/misc/crashsignal.py", line 173, in destroy_crashlogfile
    os.remove(self._crash_log_file.name)
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/qutebrowser-basedir-52n2gr5h/data/crash.log'
```
Any ideas?

I have qutebrowser-1.0.2-1 from the Arch repos and everything worked fine just a couple of hours ago (until yesterday evening).

**Comment by The-Compiler** [comment]
Hmm, this is odd! I had noticed something similar before v1.0, and fixed it by making sure the IPC socket is cleaned up before opening the subprocess and shutting down:

https://github.com/qutebrowser/qutebrowser/blob/0d1e716760add1c1e28f77b7768830a278982eb8/qutebrowser/app.py#L656-L658

The subprocess warning seems fine, as indeed the new qutebrowser process is still running when the old one exits. We should probably find a way to suppress it though.

Can you try to reproduce this with `--debug`, and show the log once you've run `:restart`, and (ideally, if you can get it) the init log of the new process?

**Comment by lahwaacz** [comment]
Sorry for the delay, I somehow lost track of this...

I was about to write that in the meanwhile there was an upgrade of the nvidia drivers which forced me to reboot the system (otherwise qutebrowser would not start due to some opengl error) and I couldn't reproduce it anymore after that. But after about 10 or 20 `:restart`s the IPC socket got missing again. I did `:debug-crash` to get to the log and the part related to IPC init is:
```
20:22:25 DEBUG    init       app:__init__:816 Initializing application...
20:22:25 DEBUG    ipc        ipc:send_to_running_instance:435 Connecting to /run/user/1000/qutebrowser/ipc-926d78cc8cc06333acfbf8f3317fffe5
20:22:25 DEBUG    ipc        ipc:send_to_running_instance:468 No existing instance present (error 2)
20:22:25 DEBUG    init       ipc:send_or_listen:497 Starting IPC server...
20:22:25 DEBUG    ipc        ipc:__init__:193 Not calling setSocketOptions
20:22:25 DEBUG    ipc        ipc:listen:204 Listening as /run/user/1000/qutebrowser/ipc-926d78cc8cc06333acfbf8f3317fffe5
```
But the `/run/user/1000/qutebrowser/ipc-926d78cc8cc06333acfbf8f3317fffe5` file did not exist!

I also had `watch -n 1 ls -lah /run/user/1000/qutebrowser/` in a terminal and the socket disappeared before the new qutebrowser window popped up, so the shutdown part probably works fine...

So it's only sporadic at the moment but still happens.

**Comment by The-Compiler** [comment]
Hmm, so you mean after the new instance started up, the file didn't exist?

**Comment by lahwaacz** [comment]
Exactly.

**Comment by The-Compiler** [comment]
That's odd, and might well be some kind of QLocalServer bug (unfortunately, I've seen a lot of bugs in Qt's network stack...).

I wonder whether it'd make sense for the old process to pass something like `--waitpid PID` to the new one, and the new one would wait using [os.waitpid](https://docs.python.org/3.5/library/os.html#os.waitpid) until the old one actually exited.

**Comment by The-Compiler** [comment]
FWIW since #3203 the old process did hang instead of exit - this probably doesn't change anything here though.

**Comment by lahwaacz** [comment]
I don't think so either, basically all posts above predate the merge of #3203. I was using the 1.0.2 release and now I just have to remember not to use 1.0.3 for testing this.

**Comment by The-Compiler** [comment]
That PR isn't in 1.0.3 - the patchlevel releases are always forked from the corresponding .0 release (in the `v1.0.x` branch in this case).
